Optimise module to filename

guibou opened the pull request:
(only latests commit are relevant, the other are from other MR which should be merged before this work is merged. Or we can rebase)

This is related to https://github.com/haskell/haskell-language-server/issues/4598.

This changes the file to module associating logic done during dependency
graph building.

Before, each time a module `Foo.Bar` is found, HLS is testing inside all
the import path for the existence of a relevant fiel.. It means that for
`i` import paths and `m` modules to locate, `m * n` filesystem
operations are done. Note also that this involves a lot of complex
string concatenation primitive to build the `FilePath`.

A module is tested for each `import` for each of the file of the
project. We also test for `boot` files, doubling the number of test.

In https://github.com/haskell/haskell-language-server/issues/4598 we have a project with `1100` modules, in more than 250 import
paths and we count more than `17000` `import` statments, resulting on
over 6 millions test for file existences. This project was blocking for
more than 3 minutes during HLS startup.

This commit changes the way this is computed:

- At startup, a `Map ModuleName FilePath` (the real type is a bit more
  involved for performance, multiples unit and boot files handling) is
  built by scanning all the import paths for files representing the
  different modules.
- Directory scanning is efficient and if import path only contains
  haskell module, this will never do more job that listing the files of
  the project.
- The lookup is now simplify a `Map` lookup.

The performance improvement is as follows:

- The number of IO operation is dramatically reduced, from multiples
  millions to a few recursive directories listing.
- A lot of the boilerplate of converting path had be removed.
- TODO: add an RTS stats before / after with number of allocations
- On my project, the graph building time is reduced from a few minutes
  to 3s.

Limitations:

- How to rebuild the `Map` if the content of one directory change?
- If one directory is filled with millions of files which are not of
  interested, performance can be damaged. TODO: add a diagnostic during
  this phase so the user can learn about this issue.

Code status:

- The `lookup` is not fully restored, especially it does not include the
  handling of home unit as well as reexport.
- The initialisation phase is cached inside a `TVar` stored as a top
  level identifier using `unsafePerformIO`. This is to be improved.

guibou committed:
refactor: Module name to filepath optimisation

This is related to https://github.com/haskell/haskell-language-server/issues/4598.

This changes the file to module associating logic done during dependency
graph building.

Before, each time an import of module `Foo.Bar` is found, HLS is testing
inside all the import path for the existence of a relevant file. It
means that for `i` import paths and `m` modules to locate, `m * n`
filesystem operations are done. Note also that this involves a lot of
complex string concatenation primitive to build the `FilePath`.

A module is tested for each `import` for each of the file of the
project. We also test for `boot` files, doubling the number of test.

In #4598 we have a project with `1100` modules, in more than 250 import
paths and we count more than `17000` `import` statments, resulting on
over 6 millions test for file existences. This project was blocking for
more than 3 minutes during HLS startup.

This commit changes the way this is computed:

- A new rule `GetModulesPaths` generates a `ModuleToFilenames` map,
  associating module name to `NormalizedFilePath`. This rules builds a
  cache of all the know module -> filepath associations, based on a scan
  of the filesystem extended with the content of the `knownTargets` and
  `VFS` in `extendModuleMapWithKnownTargets`. This rule is cached in
  `ShakeExtras` in `moduleToPathCache`.
- The lookup phase is not changed much, but instead of testing the
  knownTargets and VFS for all possibly paths, taking into account all
  possible import path, it just does an immediate lookup in the
  `ModuleToFilenames` maps.

The performance improvement is as follows:

- The number of IO operation is dramatically reduced, from multiples
  millions to a few recursive directories listing.
- A lot of the boilerplate of converting path had be removed.
- On my project, the graph building time is reduced from a few minutes
  to 3s.

A note about performance

Most users won't see the benefits of these change, but I think they
apply to everbody:

- We are still doing 1 lookup per `import` per module. But the lookup
  result is not multiples IO, so it should be faster by a large amount.
- Most project only have 1 (or a few) import paths so won't benefit as
  dramatically as me from this.

Points opens for discussions:

a) In the previous implementation, a lookup in `knownTargets` was
followed by a lookup in the VFS in case the value was not found in
`knownTargets`. We observed in https://github.com/haskell/haskell-language-server/issues/4754 that there may be a race condition when populating `knownTargets` in the context of `-boot` files, leading to sometime missing either the `.hs-boot` or the `.hs` file. The previous implementation was still working because of the fallsback on the VFS. In this new implementation, I tried to only target the VFS, however it seems that some file are not part of the VFS, this is unclear for me yet. Extending the module maps (read from the real filesystem) by both the content of the VFS and `knownTargets` give good results.
b) I'm not confident about the implementation of reexported modules,
especially in multiple home unit context, but tests are greens.
c) Cache invalidation. I had not fully understood the logic of the build
graph and cache invalidation. The current status is that the
`GetModulesPaths` rule is always rerun for each file, but the content is
manually cached inside `ShakeExtras` and never invalidated. As a result,
we cannot observe any change inside the import paths directories, and
changes on `knownTargets` and the VFS are also ignored.

guibou commented:
I'm running this code in production on our large codebase for the 3rd day.

a) My users are super happy, a few quotes:

![image](https://github.com/user-attachments/assets/6dfbccdf-5448-4d02-bc4b-0debc1b2b09c)

(it was more than 5 minutes before ;)

![image](https://github.com/user-attachments/assets/439e7486-7296-4441-8ac2-45c1670aecae)

![image](https://github.com/user-attachments/assets/7dd8ab59-0ed9-4ab8-ac5e-31f3f2635147)

Ok, that's all for the shameless plug ;)

I think I observed a few reload / performance issues which are not related and would definitely do some followup issues. I'm investigating (this is difficult to describe right now. Looks like HLS is not able to give diagnostic back if it is typechecking a file which depends on the file I just modified and if this typechecking takes time).

I also observed a surprising behavior, I need to investigate if that's linked to my MR or to something else:

- I had started HLS in a context where one of the dependencies (from the package db) was version X (which actually exposes a reexported symbol. e.g. `Database.Selda` which reexports `Text`).
- I then had changed my working copy to a version of my code which uses a new version of `Database.Selda`, without the reexport of `Text`, and my code was updated to `import Data.Text (Text)` everywhere it was previously using the reexport.
- From my editor, I *restarted* `hls`
- Warnings started to appear in HLS about redundant import for `import Data.Text (Text)` (which was true with the previous version of Selda). Note that the warning appeared on files which are NOT open in the editor, but HLS got notified about the change and reevaluated them.
- Opening the file in the editor, the warning disappears.

It looks like a multi-home issue, see. https://github.com/haskell/haskell-language-server/issues/3237, which would make sense considering that I had done nothing to fix multiple home. But this is still rather surprising because HLS was supposed to restart and hence should have completely ignored the old selda version. Maybe it picked some information from the cache.

guibou commented:
> We handle FOI(File of interest) and none-FOI differently.
See getModIfaceRule. This might be causing what you see @guibou

@soulomoon Thank you. This is indeed interesting. I will confirm that the bug I'm observing is not linked to my change and open an issue. This is not a dramatic issue (e.g. it is not everyday that you restart the editor with breaking changes in the packages from the package database), but it is not convenient and can be confusing.

guibou commented:
I've found a bug, when looking for module in the import path, an exception is raised if the import path is empty.

guibou commented:
> I've found a bug, when looking for module in the import path, an exception is raised if the import path is empty.

Fixed in latest commit.

guibou commented:
Improved the support for currently edit files but not yet saved to the disk. The implementation will be improved for performance but that's a good first draft.

This fixs a lot of unit test, so it is good.

I do have a few (around 20) tests faliling in ghcide-tests, but otherwise, most of the tests are ok.

Still todo:

- [ ] remove the `unsafePerformIO`
- [ ] Handle multi home units (some tests are failing, that's a good news)
- [ ] Handle reexport (same, some tests are failing, that's a good news)
- [ ] Handle reloading of the cache if files are added / removed

guibou commented:
Latest push should fix hlint and actually remove the `unsafePerformIO`: the cache is now stored in the `ShakeExtras`.

I have 3 (2 are flaky) remaining issues with `-boot` tests:

```
    cyclic module dependency with hs-boot:                             FAIL (0.63s)
      ghcide-test/exe/DiagnosticTests.hs:246:
      Got diagnostics for Uri {getUri = "file:///tmp/nix-shell.1Fwv9H/hls-test-root/extra-dir-80836663907422/ModuleA.hs"} but only expected diagnostics for [NormalizedUri 5530269853045372639 "
file:///tmp/nix-shell.1Fwv9H/hls-test-root/extra-dir-80836663907422/ModuleB.hs"] got [Diagnostic {_range = Range {_start = Position {_line = 1, _character = 22}, _end = Position {_line = 1, _c
haracter = 29}}, _severity = Just DiagnosticSeverity_Error, _code = Nothing, _codeDescription = Nothing, _source = Just "not found", _message = "Could not find module \8216ModuleB\8217.\nIt is
 not a module in the current program, or in any known package.", _tags = Nothing, _relatedInformation = Just [DiagnosticRelatedInformation {_location = Location {_uri = Uri {getUri = "file:///
tmp/nix-shell.1Fwv9H/hls-test-root/extra-dir-80836663907422/ModuleA.hs"}, _range = Range {_start = Position {_line = 1, _character = 22}, _end = Position {_line = 1, _character = 29}}}, _messa
ge = "GetLocatedImports"}], _data_ = Nothing}]
      Use -p '/cyclic module dependency with hs-boot/' to rerun this test only.
    bidirectional module dependency with hs-boot:                      FAIL (0.63s)
      ghcide-test/exe/DiagnosticTests.hs:269:
      Got diagnostics for Uri {getUri = "file:///tmp/nix-shell.1Fwv9H/hls-test-root/extra-dir-80836663907424/ModuleA.hs"} but only expected diagnostics for [NormalizedUri 5396647782368786105 "
file:///tmp/nix-shell.1Fwv9H/hls-test-root/extra-dir-80836663907424/ModuleB.hs"] got [Diagnostic {_range = Range {_start = Position {_line = 1, _character = 22}, _end = Position {_line = 1, _c
haracter = 29}}, _severity = Just DiagnosticSeverity_Error, _code = Nothing, _codeDescription = Nothing, _source = Just "not found", _message = "Could not find module \8216ModuleB\8217.\nIt is
 not a module in the current program, or in any known package.", _tags = Nothing, _relatedInformation = Just [DiagnosticRelatedInformation {_location = Location {_uri = Uri {getUri = "file:///
tmp/nix-shell.1Fwv9H/hls-test-root/extra-dir-80836663907424/ModuleA.hs"}, _range = Range {_start = Position {_line = 1, _character = 22}, _end = Position {_line = 1, _character = 29}}}, _messa
ge = "GetLocatedImports"}], _data_ = Nothing}]
      Use -p '/bidirectional module dependency with hs-boot/' to rerun this test only.
    correct reference used with hs-boot:                               FAIL (0.84s)
      ghcide-test/exe/DiagnosticTests.hs:295:
      Got diagnostics for Uri {getUri = "file:///tmp/nix-shell.1Fwv9H/hls-test-root/extra-dir-80836663907426/ModuleB.hs"} but only expected diagnostics for [NormalizedUri 4327465349865231065 "
file:///tmp/nix-shell.1Fwv9H/hls-test-root/extra-dir-80836663907426/ModuleC.hs"] got [Diagnostic {_range = Range {_start = Position {_line = 1, _character = 22}, _end = Position {_line = 1, _c
haracter = 29}}, _severity = Just DiagnosticSeverity_Error, _code = Nothing, _codeDescription = Nothing, _source = Just "not found", _message = "Could not find module \8216ModuleA\8217.\nIt is
 not a module in the current program, or in any known package.", _tags = Nothing, _relatedInformation = Just [DiagnosticRelatedInformation {_location = Location {_uri = Uri {getUri = "file:///
tmp/nix-shell.1Fwv9H/hls-test-root/extra-dir-80836663907426/ModuleB.hs"}, _range = Range {_start = Position {_line = 1, _character = 22}, _end = Position {_line = 1, _character = 29}}}, _messa
ge = "GetLocatedImports"}], _data_ = Nothing}]
      Use -p '/correct reference used with hs-boot/' to rerun this test only.
      ```

soulomoon commented on ghcide/session-loader/Development/IDE/Session.hs:
```diff
@@ -476,8 +476,9 @@ loadSessionWithOptions recorder SessionLoadingOptions{..} rootDir que = do
                 -- and also not find 'TargetModule Foo'.
-                fs <- filterM (IO.doesFileExist . fromNormalizedFilePath) targetLocations
-                pure $ map (\fp -> (TargetFile fp, Set.singleton fp)) (nubOrd (f:fs))
+                pure $ do
```
If we want to extendKnownTargets, we should do so in the `restart thread`, see Note [Serializing runs in separate thread] . Otherwise we risk of being unstablize the build system. 
For the .hs or .hs-boot mess, one of the possible solution, we can update the sessionLoader to invalidate related cache and consult the cradle again if the missing .hs or .hs-boot  is added.

guibou commented:
I think I had cleaned the module map logic initialisation:

a) Now only computed once, when a new file is added / removed. The performances are back and it also handle if a new file is added to the `KnownTarget`.
b) Still need to investigate the `-boot` flaky tests.
c) Still need to investagate / test for the reexport logic. Will do that next.

guibou commented:
I'm doing another tentative at finishing this.

Many things are unclear for me.

@soulomoon, you suggested to hook/have a look at `SMethod_TextDocumentDidOpen`  and `restart thread`, but I do have the feeling that my writing was interpreted as if I wanted to edit `knownTargets`.

However, I do not, instead, I'm reading `knownTargets` during the `GetLocatedImports` phase in order to know which files are already known by HLS (but may not exists on the drive yet).

And I'm doing that exactly at the some location as before in `GetLocatedImports`, but after calling another rule, `GetModulesPaths` that I introduced in this MR.

```
      716:   (notSourceModules, sourceModules) <- use_ GetModulesPaths file
      717:   KnownTargets targetsMap <- useNoFile_ GetKnownTargets
```

In my previous writings, here and in https://github.com/haskell/haskell-language-server/issues/4754, I claimed that it could be a race condition in the way `knownTargets` is populated.

However, I've today found another experience:

```
i=0
while cabal run ghcide-tests -- -p '/redundant import/&&/redundant import even without warning/'
do
    i=$(( $i + 1))
    echo $i
done;
```

It usually fails after a few iterations:

```
2
ghcide
  diagnostics
    redundant import even without warning: (TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-85960977570265/ModuleB.hs",fromList [NormalizedFilePath "/tmp/nix-shell
.Vw1D6P/hls-test-root/extra-dir-85960977570265/ModuleB.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-85960977570265/ModuleB.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-85960977570
265/ModuleB.hs"])
    redundant import even without warning: FAIL (0.31s)
      ghcide-test/exe/DiagnosticTests.hs:319:
      Could not find (DiagnosticSeverity_Warning,(3,0),"Top-level binding",Just "GHC-38417",Nothing) in [Diagnostic {_range = Range {_start = Position {_line = 2, _character = 7}, _end = Posit
ion {_line = 2, _character = 14}}, _severity = Just DiagnosticSeverity_Error, _code = Nothing, _codeDescription = Nothing, _source = Just "not found", _message = "Could not find module \8216Mo
duleA\8217.\nIt is not a module in the current program, or in any known package.", _tags = Nothing, _relatedInformation = Just [DiagnosticRelatedInformation {_location = Location {_uri = Uri {
getUri = "file:///tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-85960977570265/ModuleB.hs"}, _range = Range {_start = Position {_line = 2, _character = 7}, _end = Position {_line = 2, _characte
r = 14}}}, _message = "GetLocatedImports"}], _data_ = Nothing}]

1 out of 1 tests failed (0.33s)
```

However, with the folliwng change, adding a `threadDelay` before calling the `extendModuleMapWithKnownTargets` (which does call the `GetKnownTargets` as depicted above, it does not crash so easilly, but it still fails, later, and with a timeout:

```
157
ghcide
  diagnostics
    redundant import even without warning: (TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleA.hs",fromList [NormalizedFilePath "/tmp/nix-shell
.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleA.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleA.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317
865/ModuleA.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleB.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317
865/ModuleB.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleA.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317
865/ModuleA.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleB.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317
865/ModuleB.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleA.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317
865/ModuleA.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleB.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317
865/ModuleB.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleA.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317
865/ModuleA.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleB.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317
865/ModuleB.hs"])
(TargetFile NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317865/ModuleA.hs",fromList [NormalizedFilePath "/tmp/nix-shell.Vw1D6P/hls-test-root/extra-dir-14070387317
865/ModuleA.hs"])



    redundant import even without warning: FAIL
      Exception: Timed out waiting to receive a message from the server.
      Last message received:
      {
          "jsonrpc": "2.0",
          "method": "$/progress",
          "params": {
              "token": 5,
              "value": {
                  "kind": "report",
                  "message": "3/3",
                  "percentage": 100
              }
          }
      }
      
      While handling Timed out waiting to receive a message from the server.
        | Last message received:
        | {
        |     "jsonrpc": "2.0",
        |     "method": "$/progress",
        |     "params": {
        |         "token": 5,
        |         "value": {
        |             "kind": "report",
        |             "message": "3/3",
        |             "percentage": 100
        |         }
        |     }
        | }
      
      HasCallStack backtrace:
        throwIO, called at libraries/exceptions/src/Control/Monad/Catch.hs:308:12 in exceptions-0.10.9-6f10:Control.Monad.Catch
        throwM, called at libraries/exceptions/src/Control/Monad/Catch.hs:318:7 in exceptions-0.10.9-6f10:Control.Monad.Catch
        generalBracket, called at src/Control/Exception/Safe.hs:468:16 in safe-exceptions-0.1.7.4-36e9aa35d7afb566b6972dd7fc456fbea3e84f9f9f58420a5170817b50a55c4a:Control.Exception.Safe
        finally, called at src/Test/Hls.hs:572:67 in hls-test-utils-2.14.0.0-inplace:Test.Hls

1 out of 1 tests failed (60.78s)
```

That's interesting, because it shows that:

a) The issue is clearly a race condition, adding 200ms of thread delay seems to fix the problem (e.g. it usually fails in less than 5 iterations, and this addition get no failure in 157 iteration).
b) There maybe another race condition somewhere else which generates a deadlock (after 157 iterations).

I'll try to understand if the race condition I clearly reproduced here is linked to a mistake on my side (highly probable) OR if it was already existing, but the previous code was robust to it.

As I described in https://github.com/haskell/haskell-language-server/issues/4754, I'm convinced that there is a race condition which depending on scheduling, the content of `GetKnownTargets` is different. However the previous code had the `getTargetFor` logic, which coupled with (removed in this MR) `locateModuleFile` and `getFileExists` (which uses the VFS and is something I completely ignored for now)... WAIT!

Stop.

I've got it.

```haskell
 335     :         let getTargetFor modName nfp
 336     :                 | Just (TargetFile nfp') <- HM.lookupKey (TargetFile nfp) targets = do
 337     :                     -- reuse the existing NormalizedFilePath in order to maximize sharing
 338     :                     itExists <- getFileExists nfp'
 339     :                     return $ if itExists then Just nfp' else Nothing
 340     :                 | Just tt <- HM.lookup (TargetModule modName) targets = do
 341     :                     -- reuse the existing NormalizedFilePath in order to maximize sharing
 342     :                     let nfp' = fromMaybe nfp $ HashSet.lookupElement nfp tt
 343     :                     itExists <- getFileExists nfp'
 344     :                     return $ if itExists then Just nfp' else Nothing
 345     :                 | otherwise = do
 346     :                     itExists <- getFileExists nfp
 347  344:                     return $ if itExists then Just nfp else Nothing
 ```
 
 The previous code was calling `getTargetFor` for each path it had manually created inside `locateModuleFile`. In the context I care about, this map is populated with `TargetFile` and the `TargetFile` may not have been correctly populated for my `-boot` file. HENCE the two initial cases are ignored and the `otherwise` matchs AND queries the VFS for the file.
 
So:

a) I think I can confirm that there is an issue in the way `knowTargets` is populated
b) However, in the context of `TargetFile`, the third case skips the lookup in `knownTargets` and instead do a lookup in the VFS and is happy. Hence this race condition had no impact on the previous implementation. Howveer my implementation treats a missing entry in knownTarget as an error, hence fails.
c) It is unclear for my why the `threadDelay` seems to help here. It will be for another investigation.

guibou commented:
Actually, I'm wondering if instead of using `knowTargets` I could not use the vfs map `vfsVar`. I'll have a look.

guibou commented:
Tentative with using the `VFS` instead of `knownTargets` pushed, test are running, but it immediatly fixs the issue on the few small tests I ran here.

This is not a version ready for review yet (e.g. the merge between the module cache and the VFS is done on every lookup for module, so that's still a process `O(k * m)` (`k` the number of entries in the vfs and `m` the number of import paths), but the fix for that is easier, and first I want to validate that tests are OK without.

guibou commented:
I'm lost. The latest attempt, using the VFS instead of known targets, seems to works reliably on macos, but not on linux/window.

<img width="271" height="599" alt="image" src="https://github.com/user-attachments/assets/eff36268-1239-426b-bead-8bc34926877d" />

Pushing an attempt which merges the result of VFS and knownTargets, this is crazy, but we'll clean later during review once we'll have something reliable.

guibou commented:
The ugly hack which uses the merge of VFS and known target content seems to work (tests 100% passes locally, and were just close to 100% green also in CI)

I've now pushed a change which move the costly module -> filepath extension from known_targets / vfs to the cached location. Hopefully CI will be green tomorrow (that's already too late for me).

Next steps:

- [x] a) Implement the reexport part (But add tests for that, it is obvious at that point that this is not tested, this part of the code is now full of "variable not used" but no tests are failing. *edit*: I was not able to write a test for reexport, so instead I tried to copy the previous logic.
- [ ] b) Clarify the cache invalidation logic (I will ask explicit questions)
- [X] c) There is one todo linked to a possible optimisation (on `O(n) ` lookup on the import list for ALL `import`, which can be turned into a cached set. *edit* => Not related to this MR, the lookup is actually `O(n)` where `n` is the number of loaded unit. Definitively a possible optimisation if the project contains a lot of home module, but unlinked to the current MR.
- d) Clarify this VFS + known_target thing

guibou commented:
The latest push is there check that CI is still OK. if green, I'll:

a) Squash everything
b) Do a pass of cleanup to remove the elements not relevant to this MR, add a few comments
c) Submit for review with explicit point of focus (the cache invalidation logic, or actually the absence of it, and the weird hack between `knownTargets` and `VFS`)

guibou commented:
I've cleaned the MR and I'm ready for review.

See the commit message for guidance about the different changes. I'll also add two inline comments in the review to ask two questions related to:

a) Cache invalidation
b) The "hack" with `knownTargets` and `VFS`

guibou commented on ghcide/src/Development/IDE/Core/Rules.hs:
```diff
@@ -659,6 +662,121 @@ mkLevelEdges ms dep_node_keys = concatMap (\nk -> map (\lvl -> mkModuleEdge lvl
+  ShakeExtras{moduleToPathCache} <- getShakeExtras
+
+  cache <- liftIO (readTVarIO moduleToPathCache)
+  case Map.lookup (envUnique env_eq) cache of
```
This is the cache logic which is improper.

I would like the rule `GetModulesPaths` to be invalidated when:

- `GetKnownTargets` is invalidated (for this session). It is easy to do that by moving the call to `extendModuleMapWithKnownTargets` outside of the cache lookup here, but this operation is too costly to be redone for each file of the session.
- The cached value to be shared for all the file in the same session. I had not been able to implement that and this is why I've manually cached that inside `ShakeExtras`

So roughly, if there is a solution to cache `GetModulesPaths` for all files of the session and invalidate it when `GetKnownTargets` changes, it could solve most of my problems.

guibou commented on ghcide/src/Development/IDE/Core/Rules.hs:
```diff
@@ -659,6 +662,121 @@ mkLevelEdges ms dep_node_keys = concatMap (\nk -> map (\lvl -> mkModuleEdge lvl
+  let pathsFromTargetsMap = HS.toList $ mconcat (HM.elems knownTargets)
+  let pathsFromVFS = catMaybes $ map uriToNormalizedFilePath (Map.keys vfs)
+
+  -- This is the list of all paths known by HLS right now.
```
This is a place which is a bit fuzzy. If you look at the original `getTargetFor` (Removed elsewhere in this diff), it was actually taking into account the values from `knownTargets` and from the VFS, so it is coherent to use both, but it seems a bit overkill.

guibou commented:
This tests:

```
ghcide
  diagnostics
    lower-case drive: FAIL (0.11s)
```

Fails in the CI, it was not failing this morning. I've broken something while cleaning up, I'm investigating, but I don't think this blocks the overall review process.

guibou commented:
The regression should be fixed now.

soulomoon reviewed (commented):
For moduleToPathCache to sync with KnownTargets, one idea to put the moduleToPathCache TVar into the KnownTargets data strucutre. Whenever KnownTargets is rebuilt, it rebuilt. @guibou

fendor commented on ghcide/src/Development/IDE/Import/FindImports.hs:
```diff
@@ -66,85 +67,92 @@ modSummaryToArtifactsLocation nfp ms = ArtifactsLocation nfp (ms_location <$> ms
+  -- | "normal" files (e.g. @.hs@)
+  moduleMap       :: Map ModuleName (UnitId, NormalizedFilePath),
+  -- | "boot" files (e.g. @.hs-boot@)
+  moduleMapSource :: Map ModuleName (UnitId, NormalizedFilePath)
```
`ModuleName`s aren't unique across a project, iirc. I could easily have a project with the same module names multiple times. The values would need to be a list in this case?

guibou replied:
I'm not sure I understand in whcih context a `ModuleName` may not be unique. If it is not unique, then there is a conflict, no? My understanding is that different import paths will overrride in import path order:

```
guillaume@gecko ~/test_import_paths> ghci -isrcA -isrcB
GHCi, version 9.10.3: https://www.haskell.org/ghc/  :? for help
ghci> :l C
[1 of 1] Compiling C                ( srcA/C.hs, interpreted )
Ok, one module loaded.
ghci> c
"A"
ghci> 
ghci> 
Leaving GHCi.
guillaume@gecko ~/test_import_paths> ghci -isrcB -isrcA
GHCi, version 9.10.3: https://www.haskell.org/ghc/  :? for help
ghci> :l C
[1 of 1] Compiling C                ( srcB/C.hs, interpreted )
Ok, one module loaded.
ghci> c
"B"
ghci> 
ghci> 
Leaving GHCi.
guillaume@gecko ~/test_import_paths> cat srcA/C.hs
module C where

c = "A"
guillaume@gecko ~/test_import_paths> cat srcB/C.hs
module C where

c = "B"
```

fendor replied:
The issue is with multiple components. Imagine this:

```
.
├── cabal.project
├── CHANGELOG.md
├── foo
│   ├── CHANGELOG.md
│   ├── foo.cabal
│   └── src
│       └── MyLib.hs
├── hie.yaml
├── mult-with-package-imports.cabal
└── src
    ├── Lib.hs
    └── MyLib.hs
```

`MyLib` occurs in two home units, and the `ModuleName` is `MyLib` for both home units. In my understanding, `ModuleToFilenames` is global, so would overwrite one of the `MyLib` with the other.

Funnily enough, even though I think this should mean that `moduleMap` may be shadowing, I can't get it actually to fail in a local checkout.

guibou replied:
Indeed, you are right.

I think you cannot reproduce because in this context, there is two `cabal`, hence there will be two sessions, and the `ModuleMap` is global to one `envUnique`, so we are fine.

In order to reproduce. I think you need to have the same "session" with multiple `-unit`. However, what's the expected behavior of GHC in this context?

It seems to also shadow:

```
guillaume@gecko ~/test_import_paths> ghc unit @unitA.flags -unit @unitB.flags -unit @unitMain.flags Main.hs
[1 of 4] Compiling C                ( srcA/C.hs, srcA/C.o )[b-0.0.0.0]
[2 of 4] Compiling C                ( srcA/C.hs, srcA/C.o )[main]
[3 of 4] Compiling Main             ( Main.hs, Main.o )[main]
[4 of 4] Linking Main[main]
guillaume@gecko ~/test_import_paths> ./Main 
"A"
guillaume@gecko ~/test_import_paths> ghc unit @unitB.flags -unit @unitA.flags -unit @unitMain.flags Main.hs
[1 of 4] Compiling C                ( srcB/C.hs, srcB/C.o )[a-0.0.0.0]
[2 of 4] Compiling C                ( srcB/C.hs, srcB/C.o )[main]
[4 of 4] Linking Main[main] [Objects changed]
^[[A⏎                                                                                                                                                                                           guillaume@gecko ~/test_import_paths> ./Main
"B"
guillaume@gecko ~/test_import_paths> cat unitA.flags unitB.flags unitMain.flags 
-this-unit-id a-0.0.0.0
-isrcA
C
-this-unit-id b-0.0.0.0
-isrcB
C
-this-unit-id main
-i.
Main
guillaume@gecko ~/test_import_paths> cat srcA/C.hs
module C where

c = "A"
guillaume@gecko ~/test_import_paths> cat srcB/C.hs
module C where

c = "B"
```

See how the first unit seem to take precedence. So shadowing seems the "right behavior". However the order in my implementation may not be correct.

Maybe this could be added as a unit test.

guibou replied:
Wait, I've made a mistake (see the missing `-` in front of `-unit`.)

guibou replied:
I'm a bit lost at creating a repro here (I actually do not understand why using ghc 9.10 the `unit` here without the `-` does "work". With the `-`, it fails building `Main.hs` because it does not find `C`, and adding `-package-id a-0.0.0.0` or anything like that does not fix the problem (it complains about impossible to localise package).

fendor replied:
Dont worry about it too much, I am on it!

fendor replied:
GHC will do the right thing, it will decide which ModuleName to use based on the home unit dependency graph. Not shadow the ambigous ModuleName. that's also one of the reasons, it is always using `Module` nowadays, as it has the UnitId attached.

guibou replied:
Ok.

Keep in mind that the key of this map is used based on the `import Xxx.Yyyy.Zzzz` module name, so unless I'm wrong, that's a `ModuleName` which does not explictily state what is the `UnitId`, unless resolved later.

fendor replied:
You are correct, but we need to handle this source of ambiguation. So the map key is right, but I think the value should be a list of something similar for the case, where multiple modules have the same name but come from different units.

fendor commented:
>   -- TODO: we are taking/dropping extension, this could be factorized to save a few cpu cycles ;)
>  -- TODO: do acceptedextensions needs to be a set ? or a vector?

I benchmarked.

<img width="2388" height="1112" alt="image" src="https://github.com/user-attachments/assets/9abf82bf-fd4e-4405-93d0-071ed81590d0" />

[index.html](https://github.com/user-attachments/files/28681540/index.html)

https://gist.github.com/fendor/51a41bd4eed8d41fda79f2c36e0e86a7

Verdict, if we expect happy path to be most common, then `Vector` is the fastest, otherwise, Set seems to be the best allrounder.
I assume, the happy case `.hs`, is the most common, so I would use `Vector`.

fendor commented:
> For moduleToPathCache to sync with KnownTargets, one idea to put the moduleToPathCache TVar into the KnownTargets data strucutre. Whenever KnownTargets is rebuilt, it rebuilt. @guibou

the rule already depends on `KnownTargets`, wouldn't that mean it is rebuilt when the `KnownTargets` change any way?

soulomoon commented:
> > For moduleToPathCache to sync with KnownTargets, one idea to put the moduleToPathCache TVar into the KnownTargets data strucutre. Whenever KnownTargets is rebuilt, it rebuilt. @guibou
> 
> the rule already depends on `KnownTargets`, wouldn't that mean it is rebuilt when the `KnownTargets` change any way?

Nop, from the code, the the cache is stored externally in the shake extra . The rule only `extendModuleMapWithKnownTargets` for specific path if cache missed, but won't do any invalidation when the cache hit.

soulomoon commented:
perhaps stop storing externally in the shake extra and let the shake database handle it as rule result cache ?

fendor commented:
For anyone interested, this PR has a considerable overlap with https://gitlab.haskell.org/ghc/ghc/-/merge_requests/15888 which implements a similar optimisation in GHC itself. Which I think is quite interesting!

wz1000 committed:
cleanup

wz1000 committed:
Compute the module path map once per session

The map replaces a file system probe per candidate location per import, so
it has to be built once per session. We use the technique described in
Note [Session representatives] instead of out of band state.

wz1000 committed:
Respect unit visibility and import path order in module lookup

Which file an import resolves to depends on the unit importing it.
A unit sees only itself and its dependencies, two units may both provide a
module, and within a unit the first import directory to provide it wins.

Map a module name to the units providing it, and let the caller provide an
ordered list of units to search, as is done in the GHC finder. A file open
only in the editor is picked over the file system.

wz1000 committed:
Track the files the client reports in the known targets

The module map is derived from the file system, so it has to be told when the
file system changes, and under the closed world assumption the client tells us
everything we need to know about the state of the world.

We also have to mark it dirty, hls-graph only re-runs a node when it is marked
dirty.

See Note [Tombstones] and Note [Files that are not targets].

wz1000 committed:
Keep source spans of SOURCE imports on GHC 9.13+

wz1000 committed:
Do not let GHC search the file system for home modules

A miss in the finder cache returned Nothing, which means "not cached" to GHC,
so its finder went looking in the filesystem again for a module we had already
decided about. Poison the hook to answer InstalledNotFound instead, see
Note [Home modules are resolved by HLS].

This is slightly risky: GHC's search used to mask bugs in the HLS finding
logic, and now a module HLS fails to resolve is simply not found.

Only from GHC 9.11, where we have the finder cache hooks.

wz1000 committed:
Match GHC when a home unit reexports or hides a module

Before this we looked at every unit's files first and only then at every unit's
reexports. So if unit-a reexported module Foo, and unit-b provided module Foo,
we would pick unit-b:Foo, whereas GHC would pick unit-a:Foo.

Fix this.

wz1000 committed:
Resolve SOURCE imports through the source file, as GHC does

GHC's finder never looks for a boot file. It searches for the main hs file,
and the boot file is only allowed to be next to the main, non SOURCE
file, not at some arbitrary other import path.

wz1000 committed:
Scan import directories one directory at a time

Make the catch more granualar so that a single exception in a directory we
don't have permissions to list doesn't hide all the other modules we
have already found.

Also handle symlink loops

wz1000 committed:
Decide what lies below an import directory with makeRelative

We compared path components exactly, which is wrong on Windows. makeRelative
compares them the way the platform does, so c:/home/bob is below C:\Home.

wz1000 committed:
Remove a diagnostic race from typecheck-all-parents-of-interest

wz1000 committed:
Load modules the build tool has not been told about

Writing a module and adding it to the cabal file are two steps, so we
constantly see files that exist and are already imported but are in no
component. We gave such a file no session at all, and the modules importing it
then failed to typecheck without reporting anything.

Load it with the options of the component whose import path it is below, and
warn about it, as GHC does. See Note [Modules the build tool has not been told
about].

fendor committed:
Use VirtualFileTree for test setup specifications

fendor commented on ghcide/src/Development/IDE/LSP/Notifications.hs:
```diff
@@ -115,10 +125,17 @@ descriptor recorder plId = (defaultPluginDescriptor plId desc) { pluginNotificat
+            exts <- allExtensions <$> getIdeOptionsIO (shakeExtras ide)
+            let sourceFiles c =
+                  [ nfp | (nfp, c') <- fileEvents', c' == c
+                  , takeExtension (fromNormalizedFilePath nfp) `elem` map ('.':) exts ]
```
Why do we need to do this? To record whether haskell files were created or deleted? This feels a little ad-hoc here

wz1000 replied:
Yes. The notifications are what notify us about the state of the world changing, so we don't have keep looking at the filesystem every time. These are the files we care care about for import resolution.

In fact, we can't really look at the filesystem everytime - hls-graph works in reverse mode, so it starts with the set of dirty keys, then the things that depend on them, and so on.

If wanted to look at the filesystem, we would have to make a new always rerun rule, then mark it as dirty on notifications so it would actually rerun, and then look at the filesystem everytime that rule changes.

fendor replied:
I think I would prefer it if this logic didn't live in Notifications.hs, but it is fine for now.

fendor commented on ghcide/src/Development/IDE/LSP/Notifications.hs:
```diff
@@ -94,9 +100,13 @@ descriptor recorder plId = (defaultPluginDescriptor plId desc) { pluginNotificat
         \ide vfs _ (DidCloseTextDocumentParams TextDocumentIdentifier{_uri}) -> liftIO $ do
           whenUriFile _uri $ \file -> do
               let msg = "Closed text document: " <> getUri _uri
+              -- A file that was only ever open in the editor stops existing
+              -- when it is closed
+              onDisk <- doesFileExist (fromNormalizedFilePath file)
```
Does this work? I am not whether this simply won't race and we should rather rely on SMethod_WorkspaceDidChangeWatchedFiles notifications

wz1000 replied:
Do we get `SMethod_WorkspaceDidChangeWatchedFiles` notifications for files that weren't on the filesystem? I'm not sure, I was assuming not.

fendor replied:
Right, that's true... I guess I didn't fully grasp what these changes entailed, thank you for the clarification!

fendor commented:
I am rebase merging these commits as all of them can be reviewed standalone and bring individual improvments.
